File: 链蒸蒸简单/按身高排序.java
package 链蒸蒸简单;

import java.util.Arrays;
import java.util.PriorityQueue;

public class 按身高排序 {
    public String[] sortPeople(String[] names, int[] heights) {
        var pq = new PriorityQueue<Integer>((a, b) -> {
            return heights[b] - heights[a];
        });
        for (int i = 0; i < names.length; i++) {
            pq.add(i);
        }
        var res = new String[names.length];
        int ptr = 0;
        while (!pq.isEmpty()) {
            int index = pq.poll();
            res[ptr] = names[index];
            ptr++;
        }
        return res;
    }
}

class dedre {
    public String[] sortPeople(String[] names, int[] heights) {
        int n = names.length;
        String[] res = new String[n];
        Integer[] idx = new Integer[n];
        for (int i = 0; i < n; i++) idx[i] = i;
        Arrays.sort(idx, (a, b) -> heights[b] - heights[a]);
        for (int i = 0; i < n; i++) res[i] = names[idx[i]];
        return res;
    }
}

File: 链蒸蒸简单/按身高排序.c
#include  <stdlib.h>

int *HEIGHT;

int compare(const void *a, const void *b) {
    int aa = *(int *) a, bb = *(int *) b;
    return HEIGHT[bb] - HEIGHT[aa];
}

char **sortPeople(char **names, int namesSize, int *heights, int heightsSize, int *returnSize) {
    HEIGHT = heights;
    int idx[namesSize];
    for (int i = 0; i < namesSize; i++) idx[i] = i;
    qsort(idx, namesSize, sizeof(int), compare);
    char **res = (char **) malloc(namesSize * sizeof(char *));
    for (int i = 0; i < namesSize; i++) res[i] = names[idx[i]];
    *returnSize = namesSize;
    return res;
}